### PR TITLE
net/tcp: add Selective-ACK support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/tcp: add Selective-ACK support
net/tcp: d_appdata should remove the tcp specific option field
net/tcp: add out-of-order segment support
mm/iob: add qentry private data support

Reference:
https://datatracker.ietf.org/doc/html/rfc2018

## Impact

TCP accurate fast retransmission and recovery

## Testing

HTTP audio streaming

### CODE REVIEW

**Review by antmerlino — CHANGES_REQUESTED** [review]
We should split this PR into 2. The IOB change should be separate from the TCP changes.

**Comment by anchao** [comment]
> We should split this PR into 2. The IOB change should be separate from the TCP changes.

OK, I split the iob part to https://github.com/apache/incubator-nuttx/pull/4211

File: net/tcp/tcp_input.c [hunk 1]
```diff
@@ -565,9 +836,24 @@ static void tcp_input(FAR struct net_driver_s *dev, uint8_t domain,
             }
           else
             {
-              /* We never queue out-of-order segments. */
+#ifdef CONFIG_NET_TCP_OUT_OF_ORDER_QUEUE
+              dev->d_len += hdrlen;
+              tcp_input_ofo(dev, conn, iplen);
+
+#  ifdef CONFIG_NET_TCP_SELECTIVE_ACK
+              if (conn->flags | TCP_SACK)
```
Comment (yamt) [inline net/tcp/tcp_input.c:844]: i guess you meant something different here.

**Review by yamt — CHANGES_REQUESTED** [review]
i think this PR have too many things for a single review/merge/test cycle.
i'd suggest to create a PR with CONFIG_NET_TCP_OUT_OF_ORDER_QUEUE alone.

**Comment by anchao** [comment]
Hi @yamt @antmerlino ,

I refactored the related implementation of out-of-order and sack based on the latest dev, let us move https://github.com/apache/nuttx/pull/8062 to review this feature further

**Pull request closed without merging**